[B+C] Stores mob spawn reason inside the mob. Adds BUKKIT-2598

meiskam opened the pull request:
**The Issue**:

Currently, mob spawn reason is only available inside of a CreatureSpawnEvent.

**Justification for this PR**:

Multiple plugins save this data in maps for various reasons, which is usually lost after a server restart.  This request centralizes the data to one location, and is persistent over restarts.

**PR Breakdown**:

On creature spawn, the reason is stored in an internal variable inside the living entity.  On map save/load it is stored inside the mob's NBT tag Bukkit.SpawnReason.

**Testing Results and Materials**:

On mob damage, their spawn reason is messaged to the damager.  Plugin source is inside jar.
CraftBukkit: http://dropbox.meiskam.shininet.org/minecraft/bukkit/test/MobSpawnReason/craftbukkit-b3006-MobSpawnReason.jar
Plugin: http://dropbox.meiskam.shininet.org/minecraft/bukkit/test/MobSpawnReason/MobSpawnReasonTest-1.2.jar

**Relevant PR**:

CB-1152 - https://github.com/Bukkit/CraftBukkit/pull/1152

**JIRA Ticket**:

BUKKIT-2598 - https://bukkit.atlassian.net/browse/BUKKIT-2598

ST-DDT commented:
Wouldn't it be better to allow accessing the NBTTags directly via Bukkit and allow storing of any kind of data?

EdGruberman commented:
@ST-DDT: No.  NBT is an implementation specific storage mechanism that can not be guaranteed in the future.

@meiskam: I don't see the use case for this outside of convenience.  And then the impact is CraftBukkit has to store additional information in the region files accordingly.  Plugins can simply store this information if they need to themselves.

riking commented:
Additionally, plugins can store their data caches inside of the world folder so that it is picked up by a backup. The issue of not being restored from a backup is really a plugin-side issue.

meiskam commented:
Updated per @EdGruberman's suggestion (on JIRA): if a mob is naturally spawned, it won't require any data storage.  This will majorly reduce any increase in file size caused by this pull.

gravitylow commented on src/main/java/org/bukkit/entity/LivingEntity.java:
```diff
@@ -352,4 +353,11 @@
     public boolean isCustomNameVisible();
+
+    /**
+     * Gets the reason the living entity spawned into the world
```
Needs a period at the end of the sentence.

gravitylow commented:
Commit message needs a better explanation per https://github.com/Bukkit/Bukkit/blob/master/CONTRIBUTING.md#commit-message-example

meiskam committed:
Stores mob spawn reason inside the mob. Adds BUKKIT-2598

The reason a mob spawned is now available any time, instead of only during
the spawn event

turt2live commented:
@meiskam If you, or anyone else, is still interested in this PR please address the following concerns:
- A stronger justification for use

Thanks!
